src/datasources/email/email.datasource.ts

```diff
@@ -0,0 +1,157 @@
+import { EmailAddressDoesNotExistError } from '@/datasources/email/errors/email-address-does-not-exist.error';
+import { EmailAlreadyVerifiedError } from '@/datasources/email/errors/email-already-verified.error';
+import { InvalidVerificationCodeError } from '@/datasources/email/errors/invalid-verification-code.error';
+import { ExpiredVerificationCodeError } from '@/datasources/email/errors/expired-verification-code.error';
+import { IConfigurationService } from '@/config/configuration.service.interface';
+
+@Injectable()
+export class EmailDatasource implements IEmailDataSource {
+  private readonly verificationExpireTimeInSeconds: number;
+
+  constructor(
+    @Inject('DB_INSTANCE') private readonly sql: postgres.Sql,
+    @Inject(IConfigurationService) configurationService: IConfigurationService,
+  ) {
+    this.verificationExpireTimeInSeconds = configurationService.getOrThrow(
+      'email.verificationExpireTimeInSeconds',
+    );
+  }
+
+  private _resetVerificationCode(
+    sql: postgres.Sql,
+    emailId: number,
+  ): postgres.PendingQuery<VerificationStatus[]> {
+    // TODO proper code generation
+    const code = Math.floor(Math.random() * 100_000);
+    return sql<VerificationStatus[]>`
+        INSERT INTO emails.verification (id, verification_code)
+        VALUES (${emailId}, ${code})
+        ON CONFLICT (id) DO UPDATE SET verification_code = ${code},
+                                       sent_on           = now()
+        RETURNING *
+    `;
+  }
+
+  private getEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+  }): postgres.PendingQuery<Email[]> {
+    return this.sql<Email[]>`SELECT *
+                             FROM emails.signer_emails
+                             WHERE chain_id = ${args.chainId}
+                               and safe_address = ${args.safeAddress}
+                               and signer = ${args.signer}`;
+  }
+
+  async saveEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    emailAddress: string;
+    signer: string;
+  }): Promise<EmailVerificationCode> {
+    return await this.sql.begin(async (sql) => {
+      const [email] = await sql<Email[]>`
+          INSERT INTO emails.signer_emails (chain_id, email_address, safe_address, signer)
+          VALUES (${args.chainId}, ${args.emailAddress}, ${args.safeAddress}, ${args.signer})
+          RETURNING *
+      `;
+
+      const [verificationStatus] = await this._resetVerificationCode(
+        sql,
+        email.id,
+      );
+
+      return <EmailVerificationCode>{
+        emailAddress: email.email_address,
+        verificationCode: verificationStatus.verification_code,
+      };
+    });
+  }
+
+  async resetVerificationCode(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+  }): Promise<{ emailAddress: string; verificationCode: string }> {
+    const [email] = await this.getEmail(args);
```

Comment:
This is getting the first (ordered by insertion date) `Email` in the database, is this correct? Is this function intended to be called without specifying the email address for the reset?

Reply:
> This is getting the first (ordered by insertion date) `Email` in the database, is this correct? 

Yes.

> Is this function intended to be called without specifying the email address for the reset?

Yes. Since the combination of `signer`, `chainId` and `safeAddress` should be unique then the email doesn't need to be provided. Also this makes it easier for designing the client API – if an email was required for the code reset the frontend would need to store the email locally in order to provide it in different flows.

Reply:
I get the point regarding the client API, but I don't understand why unicity is guaranteed for these three elements. I see this in the initial migration:
```
UNIQUE (chain_id, email_address, safe_address, signer)
```
So the unique constraint includes the `email_address`, and therefore, we could have several sets of [`signer`, `chainId`, `safeAddress`] with different `email_address`.

Reply:
Ah yes this is an issue. the `email_address` should not be part of that group.